### PR TITLE
ui: fix cursor problems

### LINKED ISSUE #12800: Default Cursor highlight (guifg=bg guibg=fg) could not be used to restore from blend=100 cursor
- `nvim --version`: NVIM v0.5.0-nightly
- `vim -u DEFAULTS` (version: ) behaves differently?: Vim does not support `blend=100`
- Operating system/version: macOS Catalina 10.15.6
- Terminal name/version: iTerm2 3.3.12
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
# Requirements
:set termguicolors
:highlight InvisibleCursor gui=reverse blend=100
# Hide cursor
:set guicursor+=a:InvisibleCursor
# Restoring cursor with Cursor (guifg=bg guibg=fg) does not work
:set guicursor+=a:Cursor
# Restoring cursor with TermCursor (cterm=reverse gui=reverse) works
:set guicursor+=a:TermCursor
```

Kapture 2020-08-28 at 09 04 54

### Actual behaviour

`Cursor` highlight is not applied thus the cursor keeps invisible.

### Expected behaviour

`Cursor` highlight applied and the cursor become visible.

### Note

If I define `Cursor` highlight manually (e.g. `ctermfg=234 ctermbg=252 guifg=#161821 guibg=#c6c8d1`), `set guicursor+=a:Cursor` restore cursor thus it seems the highlight `guifg=bg guibg=fg` cause this issue.

### Related
https://github.com/neovim/neovim/issues/12796
https://github.com/lambdalisue/fern.vim/issues/177

### PR DESCRIPTION
**Pull request by erw7**
- Change to reset everything before setting `guicursor`.
- tui: Fix a case where the cursor was not displayed after hiding the   cursor and then setting it to be displayed again.

fixes #12800.

### Steps to reproduce using `nvim -u NORC`

```vim
nvim -u NORC
:set termguicolors
:hi nCursor guifg=red guibg=red
:hi iCursor guifg=green guibg=green
:hi cCursor guifg=blue guibg=blue
:set guicursor=n:block-nCursor,i:hor25-iCursor,c:ver25-cCursor
:set guicursor-=c:ver25-cCursor 
:
```

### Actual behaviour

The cursor is a blue vertical.

### Expected behaviour

The cursor should be the default color block.

- [x] Document.
- [x] Test.

### CODE REVIEW

File: src/nvim/cursor_shape.c [hunk 1]
```diff
@@ -99,11 +99,23 @@ char_u *parse_shape_opt(int what)
 
   // First round: check for errors; second round: do it for real.
   for (round = 1; round <= 2; round++) {
+    // Set the defaults, for the missing parts
+    if (round == 2 || *p_guicursor == NUL) {
+      for (idx = 0; idx < SHAPE_IDX_COUNT; idx++) {
```
Comment (justinmk) [inline src/nvim/cursor_shape.c:104]: this for loop should go in a `static clear_shap_table()` function. That avoids setting `idx` here which is fragile, also helps readability.

File: src/nvim/cursor_shape.c [hunk 2]
```diff
@@ -99,11 +99,23 @@ char_u *parse_shape_opt(int what)
 
   // First round: check for errors; second round: do it for real.
   for (round = 1; round <= 2; round++) {
+    // Set the defaults, for the missing parts
```
Comment (justinmk) [inline src/nvim/cursor_shape.c:102]: "for the missing parts" is no longer relevant, right?

File: src/nvim/cursor_shape.c [hunk 3]
```diff
@@ -99,11 +99,23 @@ char_u *parse_shape_opt(int what)
```
Comment (justinmk) [inline src/nvim/cursor_shape.c:99]: let's mention this in the docstring for `parse_shape_opt`:

```
/// Parses the 'guicursor' option, or clears `shape_table` if 'guicursor' is empty.
///
/// @param ...

**Review by justinmk — APPROVED** [review]
looks like a good idea, definitely worth testing but can probably just test it by checking 'guicursor' before/after, don't need a TUI test.

File: test/functional/ui/cursor_spec.lua [hunk 4]
```diff
@@ -286,6 +286,21 @@ describe('ui/cursor', function()
       eq(173, named.normal.blinkon)
       eq(42, named.showmatch.cell_percentage)
     end)
+
+    -- If there is no setting for guicursor, it becomes the default setting.
+    meths.set_option('guicursor', 'n:ver35-blinkwait171-blinkoff172-blinkon173-Cursor/lCursor')
+    screen:expect(function()
+      for _,m in ipairs(screen._mode_info) do
+        if m.name ~= 'normal' then
```
Comment (justinmk) [inline test/functional/ui/cursor_spec.lua:294]: should we explicitly test the "normal" case also?

Comment (erw7) [reply test/functional/ui/cursor_spec.lua:294]: I think the case where the setting exists in `guicursor` has been well tested in previous tests. Therefore, I don't think it's necessary.

File: src/nvim/cursor_shape.c [hunk 5]
```diff
@@ -77,7 +81,8 @@ Array mode_style_array(void)
   return all;
 }
 
-/// Parse the 'guicursor' option
+/// Parse the 'guicursor' option, or clears `shape_table` if 'guicursor'
```
Comment (teto) [inline src/nvim/cursor_shape.c:84]: I think doxygen allows to reference symbols via `\ref` so you could write `\ref  shape_table`.

**Review by teto — COMMENTED** [review]
@lambdalisue does that fix it for you ?

**Comment by justinmk** [comment]
will merge this today after some doc tweaks

**Pull request closed without merging**